apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -0,0 +1,606 @@
+            {block_height, block_hash},
+            0,
+            {{block_height, block_hash}, %{block: block}}
+          )
+
+        _ ->
+          state.hash_pool
+      end
+
+    Logger.info("#{__MODULE__}: fetch next from Hashpool")
+
+    case update_chain_from_pool(height_in, hash_in, hash_pool) do
+      {:error, reason} ->
+        Logger.info("#{__MODULE__}: Chain update failed: #{inspect(reason)}")
+        {:reply, {:error, reason}, %{state | hash_pool: hash_pool}}
+
+      {:ok, new_height, new_hash, []} ->
+        Logger.debug(fn -> "Got all the blocks from Hashpool" end)
+
+        # The sync might be done. Check for more blocks.
+        case Enum.find(state.sync_pool, false, fn peer -> Map.get(peer, :id) == peer_pid end) do
+          false ->
+            # Abort sync, we don't have this peer in our list anymore.
+            {:reply, {:error, :unknown_peer}, %{state | hash_pool: []}}
+
+          %{to: to} when to <= new_height ->
+            # We are done!
+            new_sync_pool =
+              Enum.reject(state.sync_pool, fn peers -> Map.get(peers, :id) == peer_pid end)
+
+            {:reply, :done, %{state | hash_pool: [], sync_pool: new_sync_pool}}
+
+          _peer ->
+            # This peer has more blocks to give. Fetch them!
+            {:reply, {:fill_pool, new_height, new_hash}, %{state | hash_pool: []}}
+        end
+
+      {:ok, new_height, new_hash, new_hash_pool} ->
+        sliced_hash_pool =
+          for {{height, hash}, %{peer: _id}} <- new_hash_pool do
+            {height, hash}
+          end
+
+        case sliced_hash_pool do
+          [] ->
+            # We have all blocks. Just insertion left
+            {:reply, {:insert, new_height, new_hash}, %{state | hash_pool: new_hash_pool}}
+
+          pick_from_hashes ->
+            # We still have blocks to fetch.
+            {_pick_height, pick_hash} = Enum.random(pick_from_hashes)
+
+            {:reply, {:fetch, new_height, new_hash, pick_hash},
+             %{state | hash_pool: new_hash_pool}}
+        end
+    end
+  end
+
+  def handle_cast({:delete_from_pool, peer_pid}, %{sync_pool: pool} = state) do
+    {:noreply, %{state | sync_pool: Enum.filter(pool, fn peer -> peer.peer != peer_pid end)}}
+  end
+
+  def handle_info({:gproc_ps_event, event, %{info: info}}, state) do
+    case event do
+      :block_created -> enqueue(:forward, %{status: :created, block: info})
+      :tx_created -> enqueue(:forward, %{status: :created, tx: info})
+      :top_changed -> enqueue(:forward, %{status: :top_changed, block: info})
+      :tx_received -> enqueue(:forward, %{status: :received, tx: info})
+    end
+
+    Jobs.dequeue(:sync_jobs)
+    {:noreply, state}
+  end
+
+  def handle_info({:DOWN, _ref, :process, pid, reason}, %{sync_pool: sync_pool} = state) do
+    Logger.info("Worker stopped with reason: #{inspect(reason)}")
+    {:noreply, %{state | sync_pool: Enum.filter(sync_pool, fn peer -> peer.peer != pid end)}}
```

Comment:
It is said that peer is String.t() (in the type definition), what exactly is it (I don't understand peer.peer != pid)? List.delete should be better here besides that. Is the process that is being monitored a sync job?

Reply:
Good catch, the peer is a pid() actually. And List.delete is a good option too.